# 1. Importing libraries

In [86]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 1.1 Importing Data

In [87]:
path = r'C:\Users\Gregor\ownCloud\CareerFoundry\10-2023 Instacart Basket Analysis'

In [88]:
# Import the orders.csv file
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [89]:
# Import the products.csv file
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 1.2 Wrangling Data

In [90]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [91]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [92]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [93]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [94]:
df_ords['order_id'].dtype

dtype('O')

In [95]:
# Import the departments.csv file
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [96]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [97]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [98]:
df_dep_t = df_dep.T

In [99]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [100]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [101]:
new_header = df_dep_t.iloc[0]

In [102]:
new_header

0    department
Name: department_id, dtype: object

In [103]:
df_dep_t_new = df_dep_t[1:]

In [104]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [105]:
df_dep_t_new.columns = new_header

In [106]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [107]:
data_dict = df_dep_t_new.to_dict('index')

In [108]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [109]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [110]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [111]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [112]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [113]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [114]:
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [115]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [116]:
# df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

# Step 2
Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [117]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [118]:
df_ords

order_id user_id eval_set  order_number  orders_day_of_week  \
0        2539329       1    prior             1                   2   
1        2398795       1    prior             2                   3   
2         473747       1    prior             3                   3   
3        2254736       1    prior             4                   4   
4         431534       1    prior             5                   4   
...          ...     ...      ...           ...                 ...   
3421078  2266710  206209    prior            10                   5   
3421079  1854736  206209    prior            11                   4   
3421080   626363  206209    prior            12                   1   
3421081  2977660  206209    prior            13                   1   
3421082   272231  206209    train            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

# Step 3
Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

Already changed 'orders_dow' to 'orders_day_of_week' in the wrangling data above.

# Step 4
Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [119]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

The busiest hour for placing orders is 10 o'clock.

# Step 5
Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [120]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [121]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

The meaning behind a value of 4 in the 'department_id' refers to 'produce'.

# Step 6
The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [122]:
df_breakfast =  df_prods[df_prods['department_id']==14]

In [123]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# Step 7
They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [124]:
df_dinner = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [125]:
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# Step 8
It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

The df_dinner dataframe contains 7650 rows.

# Step 9
Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [126]:
df_ords[df_ords['user_id']=='1']

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

# Step 10
You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [127]:
df_ords[df_ords['user_id']=='1'].describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# Step 11-13
Check the organization and structure of your notebook. Be sure to include section headings and code comments.
Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.
Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [128]:
# Exporting df_ords to 'orders_wrangled.csv' 
# df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [129]:
# Exporting df_dep_t_new to 'departments_wrangled.csv' 
# df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))

In [130]:
df_test = pd.DataFrame()

# Exercise 4.5 Data Consistency Checks

In [131]:
df_test['mix'] = ['a', 'b', 1, True]

# 4.5 Step 1
If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

In [132]:
df_test

mix
0     a
1     b
2     1
3  True

In [133]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [134]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Gregor\AppData\Local\Temp\ipykernel_15412\1786467950.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [135]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [136]:
df_test['mix'] = df_test['mix'].astype('str')

In [137]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [138]:
df_prods.shape

(49693, 5)

In [139]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [140]:
df_prods_clean.shape

(49677, 5)

In [141]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [142]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [143]:
df_prods_clean.shape

(49677, 5)

In [144]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [145]:
df_prods_clean_no_dups.shape

(49672, 5)

# Step 2
Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

    Tip: Keep an eye on min and max values!


In [146]:
df_ords.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

Most of the values seem reasonable. The count of 'days_since_prior_order' is smaller than the count of the other columns indicating a larger number of missing values in that specific column.

In [147]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum value of 99999.0 for prices looks way too large when looking at the 75 percentile with a value of 11.1.

# Step 3
Check for mixed-type data in your df_ords dataframe.

In [148]:
# For-loop to find mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\Gregor\AppData\Local\Temp\ipykernel_15412\1149698198.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_15412\1149698198.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_15412\1149698198.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_15412\1149698198.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\

There is no output from the for-loop and thus, no column in the orders dataframe has mixed data types

# Step 4
If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

No mixed-type data found.

# Step 5
Run a check for missing values in your df_ords dataframe.



In [149]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [155]:
df_ords.head(20)

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   
11  2168274       2    prior             1                   2   
12  1501582       2    prior             2                   5   
13  1901567       2    prior             3                   1   
14   738281       2    prior             4                   2   
15  1673511       2    prior             5                   3   
16  1199898       2    prior             6                   2   
17  3194192       2    prior             7                   2   
18   788338       2    prior             8                   1   
19  1718559       2    prior             9                   2   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     NaN  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0

There are 206209 missing values in the days_since_prior_order column, as expected from the df_ords.describe() output. The missing values are probably due to the fact that when a user/costumer makes a first order, there are no prior orders. Therefore , each user has for the first order number a missing value in the 'days_since_prior_order' column.

# Step 6
Address the missing values using an appropriate method.

    In a markdown cell, explain why you used your method of choice.


In [156]:
df_ords

order_id user_id eval_set  order_number  orders_day_of_week  \
0        2539329       1    prior             1                   2   
1        2398795       1    prior             2                   3   
2         473747       1    prior             3                   3   
3        2254736       1    prior             4                   4   
4         431534       1    prior             5                   4   
...          ...     ...      ...           ...                 ...   
3421078  2266710  206209    prior            10                   5   
3421079  1854736  206209    prior            11                   4   
3421080   626363  206209    prior            12                   1   
3421081  2977660  206209    prior            13                   1   
3421082   272231  206209    train            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [150]:
df_ords['days_since_prior_order'].median()

7.0

In [151]:
df_ords['days_since_prior_order'].mean()

11.114836226863012

As the number of missing values in the days_since_prior_order column correspond to the number of distinct user ids, i.e. 206209, I decided not to make any changes. Even though it looks missing, actually the data itself has value in it. I can use it to show that it's actually the user's first order made or it is a new customer who just made his first order.

## Step 7-8
Run a check for duplicate values in your df_ords data.

    In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

Address the duplicates using an appropriate method.

    In a markdown cell, explain why you used your method of choice.



In [152]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [153]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are no dupicate values in the df_ords data.

## Step 9
Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [158]:
# Exporting  df_prods_clean_no_dups to 'products_checked.csv' 
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [159]:
# Exporting  df_ords to 'orders_checked.csv' 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))